src/include/miopen/invoke_params.hpp

```diff
@@ -0,0 +1,168 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2019 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/errors.hpp>
+
+#include <memory>
+#include <typeinfo>
+#include <type_traits>
+#include <utility>
+
+namespace miopen {
+
+enum class InvokeType
+{
+    Run,
+    Evaluate,
+    AutoTune,
+};
+
+struct InvokeParams
+{
+    InvokeType type = InvokeType::Run;
+};
+
+struct AnyInvokeParams
+{
+    public:
+    AnyInvokeParams() = default;
+
+    template <class Actual>
+    AnyInvokeParams(const Actual& value) : impl(std::make_unique<Implementation<Actual>>(value))
+    {
+    }
+
+    template <class Actual,
+              class = std::enable_if_t<!std::is_same<Actual, AnyInvokeParams>::value, void>>
+    AnyInvokeParams(Actual&& value)
+        : impl(std::make_unique<Implementation<Actual>>(std::forward(value)))
+    {
+    }
+
+    AnyInvokeParams(const AnyInvokeParams& other) : impl(other.impl ? other.impl->Copy() : nullptr)
+    {
+    }
+
+    AnyInvokeParams(AnyInvokeParams&& other) noexcept : impl(std::move(other.impl)) {}
+
+    AnyInvokeParams& operator=(const AnyInvokeParams& other)
+    {
+        if(&other != this)
+            impl = other.impl ? other.impl->Copy() : nullptr;
+        return *this;
+    }
```

Comment:
The assignment operator should use copy-swap:

```cpp
    AnyInvokeParams& operator=(AnyInvokeParams other)
    {
        impl.swap(other.impl);
        return *this;
    }
```

This provides the strong exception guarantee.